x-pack/plugins/event_log/server/es/cluster_client_adapter.ts

```diff
@@ -238,41 +239,112 @@ export class ClusterClientAdapter<TDoc extends { body: AliasAny; index: string }
       });
       throw err;
     }
+    const savedObjectsQueryMust: estypes.QueryDslQueryContainer[] = [
+      {
+        term: {
+          'kibana.saved_objects.rel': {
+            value: SAVED_OBJECT_REL_PRIMARY,
+          },
+        },
+      },
+      {
+        term: {
+          'kibana.saved_objects.type': {
+            value: type,
+          },
+        },
+      },
+      // @ts-expect-error undefined is not assignable as QueryDslTermQuery value
+      namespaceQuery,
+    ];
+    if (!legacyIds) {
+      savedObjectsQueryMust.push({
+        terms: {
+          // default maximum of 65,536 terms, configurable by index.max_terms_count
+          'kibana.saved_objects.id': ids,
+        },
+      });
+    }
     const musts: estypes.QueryDslQueryContainer[] = [
       {
         nested: {
           path: 'kibana.saved_objects',
           query: {
             bool: {
-              must: [
-                {
-                  term: {
-                    'kibana.saved_objects.rel': {
-                      value: SAVED_OBJECT_REL_PRIMARY,
+              must: reject(savedObjectsQueryMust, isUndefined),
+            },
+          },
+        },
+      },
+    ];
+    if (legacyIds) {
+      musts.push({
+        bool: {
+          should: [
+            {
+              bool: {
+                must: [
+                  {
+                    nested: {
+                      path: 'kibana.saved_objects',
+                      query: {
+                        bool: {
+                          must: [
+                            {
+                              terms: {
+                                // default maximum of 65,536 terms, configurable by index.max_terms_count
+                                'kibana.saved_objects.id': legacyIds,
+                              },
+                            },
+                          ],
+                        },
+                      },
+                    },
+                  },
+                  {
+                    range: {
+                      'kibana.version': {
+                        lt: '8.0.0',
+                      },
                     },
```

Comment:
**question:** will this cover scenarios where `kibana.version` is empty? I'm thinking of documents that got written before https://github.com/elastic/kibana/issues/56771 started adding this `kibana.version` value.

Reply:
Yes, it's a good point. I think in this case we need to extend the condition for legacy Ids to include this documents too.

Reply:
I'm wondering if we even need a version check here.  I think we would, if there would ever be a collision between a legacy id and new id.  But if we **KNOW** that can't happen (for example, we use a new UUID format for shareable, and perhaps there's no chance of a collision between them, or it's very very small), can we just use the existing query, and append the legacy ids and the other ids together?  Would that even change the way this is shaped, so that event log wouldn't know anything about legacy ids - alerting would just pass the legacy ids in with the other SO ids.

I'm probably missing some other reason why we need to do this anyway :-)

Reply:
>But if we KNOW that can't happen

It can happen when using pre-defined IDs. We're trying to warn users about the problems using pre-defined IDs case as much as possible (which will cause duplicate rules after upgrade). But, it is possible to create a rule in 7.x with ID `1` and create a different rule in 8.x with ID `1`. Each would be treated separately but may share event log documents for ID `1` split by version (7.x vs 8.x).

Reply:
heh, of course, **pre-defined IDs** strike again!  thx